src/main/java/org/spongepowered/common/data/processor/data/entity/PassengerDataProcessor.java

```diff
@@ -0,0 +1,136 @@
+
+    @Override
+    public boolean supports(DataHolder dataHolder) {
+        return dataHolder instanceof net.minecraft.entity.Entity;
+    }
+
+    @SuppressWarnings("unused")
+    @Override
+    public Optional<PassengerData> from(DataHolder dataHolder) {
+        if (supports(dataHolder)) {
+            final SpongePassengerData passengerData = new SpongePassengerData();
+            final Entity vehicle = (Entity) ((net.minecraft.entity.Entity) dataHolder).ridingEntity;
+            if (vehicle != null) {
+                net.minecraft.entity.Entity currentEntity = (net.minecraft.entity.Entity) vehicle;
+                while (currentEntity != null && currentEntity.isRiding()) {
+                    currentEntity = currentEntity.ridingEntity;
+                }
+                if (currentEntity != null) {
+                    final Entity baseVehicle = (Entity) currentEntity;
+                    return Optional.<PassengerData>of(passengerData.setVehicle(vehicle).setBaseVehicle(baseVehicle));
+                } else {
+                    return Optional.absent();
+                }
+            } else {
+                return Optional.absent();
+            }
+        } else {
+            return Optional.absent();
+        }
+
+    }
+
+    @Override
+    public Optional<PassengerData> fill(DataHolder dataHolder, PassengerData manipulator, MergeFunction overlap) {
+        if (supports(dataHolder)) {
+            final PassengerData merged = checkNotNull(overlap).merge(checkNotNull(manipulator).copy(), from(dataHolder).get());
+            manipulator.set(Keys.BASE_VEHICLE, merged.baseVehicle().get()).set(Keys.VEHICLE, merged.vehicle().get());
+            return Optional.of(manipulator);
+        }
+        return Optional.absent();
+    }
+
+    @Override
+    public Optional<PassengerData> fill(DataContainer container, PassengerData passengerData) {
+        passengerData.set(Keys.BASE_VEHICLE, getData(container, Keys.BASE_VEHICLE));
+        passengerData.set(Keys.VEHICLE, getData(container, Keys.VEHICLE));
+        return Optional.of(passengerData);
+    }
+
+    @Override
+    public DataTransactionResult set(DataHolder dataHolder, PassengerData manipulator, MergeFunction function) {
+        // TODO
+        return null;
+    }
+
+    @Override
+    public Optional<ImmutablePassengerData> with(Key<? extends BaseValue<?>> key, Object value, ImmutablePassengerData immutable) {
+        // TODO: Should not return Optional.absent()
+        return Optional.absent();
+    }
+
+    @Override
+    public DataTransactionResult remove(DataHolder dataHolder) {
+        if (supports(dataHolder)) {
+            net.minecraft.entity.Entity entity = ((net.minecraft.entity.Entity) dataHolder);
+            if (entity.isRiding()) {
+                entity.mountEntity(null);
+                return DataTransactionBuilder.builder().result(DataTransactionResult.Type.SUCCESS).build();
```

Comment:
You should try to check if there was an entity the dataholder was riding already, if so, provide the "replaced" result in the transaction.